### ISSUE TITLE
job-info: store historical job data in a database

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#2863]
Peeling off discussion in https://github.com/flux-framework/flux-accounting/issues/7 as well as old issues #2093 & #2455.

Problem: the job-info job list service stores all job list data in memory forever.   This is clearly a long term problem.  At some point, some amount of job history should be purged and placed somewhere (i.e. a database) for long term retrieval.

Desired Solution: Any requests to the job info job list service for older jobs should look up that data in the database.  Whether job data is cached in memory or in the database should not matter and the request for jobs should appear seemless to a user.

  - I assume store this in the `content.backing-path` location.

Fair-Share calculations: fair share calculations also need job history information for its calculations.

Questions:

- is it a requirement that fair share calculations read directly from the database?  If an appropriate interface were written, could it be done via the job list service?
  - or would fair share calculations  need to read directly from the database for its purposes

- Should job-info periodically (or at "thresholds") purge data into the database?  Or should inactive / completed jobs be regularly written out as they complete?
  - in other worlds, is fair share calculations ok reading a reasonable approximation of completed jobs in the job history database? How up to date does it have to be?

- what "format" should the data be stored in the database.  "format" is not the best word to use see below:
  - store "raw" data in the database, for example store the jobspec.  job-info can re-calculate the job's `ntasks` or `nnodes` values after re-reading the jobspec.
    - pro: any new queryable attributes would work with older jobs
    - con: a lot more data stored
    - con: possibly not easy for fair share calculations to process?
  - store the values in job-info that are retrievable by users (i.e. `nnodes` or `ntasks`).  This info is likely useful by fair share.
    - pro: allows fair share queries easy lookups by directly accessing the database
    - con: potential data loss.  If we wish to query for new information in the future, we won't have it.  In this example, the `jobspec` is lost forever.

**Comment by cmoussa1** [comment]
> is it a requirement that fair share calculations read directly from the database? If an appropriate interface were written, could it be done via the job list service?

I could be wrong but I believe fair share calculations are done by looking at completed jobs in memory (this is what allows calculation of usage over some period of time, like the past two weeks for example). It reads directly from the database in a couple of special cases (that I know of):

1. **Machine/System Restart**: it needs to look at the most recently completed jobs to rebuild its fairshare values

2. **Association Value Updates**: if an association's limits change, then I think this would also trigger a need for direct reads from a database to rebuild its fairshare values

After those couple of weeks, those jobs get pushed to persistent storage and can be retrieved at a later date.

> Should job-info periodically (or at "thresholds") purge data into the database? Or should inactive / completed jobs be regularly written out as they complete?

I believe job-info would purge job data into its database at a certain threshold (again, I believe the time period is a couple of weeks or something). Fairshare would probably need to read the data from jobs as they complete, but this could be accomplished the job listing service, right?

**Comment by chu11** [comment]
> It reads directly from the database in a couple of special cases (that I know of):

Ok, so there was a desire for `job-info` to have a DB that could be used by both `job-info` and fairshare calculations.  But what this may ultimately come down to is the following:

- for job-info, will it be a requirement to store "raw" job data (e.g. jobspec, R, eventlog) in the database.  If so, then perhaps fairshare would get data from job-info and store data into its own database for its own purposes.

OR

- if the data is stored "raw", would fairshare be ok reading this "raw" job data and parsing it on its own?

> I believe job-info would purge job data into its database at a certain threshold (again, I believe the time period is a couple of weeks or something).

Ok, sounds good.  I primary concern was the granularity of the "dumps" was outside of the needs of fairshare.

> Fairshare would probably need to read the data from jobs as they complete, but this could be accomplished the job listing service, right?

Yup.

**Comment by cmoussa1** [comment]
> if the data is stored "raw", would fairshare be ok reading this "raw" job data and parsing it on its own?

I think this is a good strategy. That way we wouldn't need to keep multiple sets of data. It could just read data that falls into that "threshold" time period to perform its calculations.

**Comment by chu11** [comment]
I've been pondering the best steps to implement this, as getting from where the jobinfo module is today to its end goal is multiple steps away and each step non-trivial, and it's a little hard for me to currently see what the end result will look like code wise.

But I'm thinking:

1. save off all inactive data to database.  That way there is data for @cmoussa1 to use.  We won't purge any data from memory at this point.
2. purge data from memory in some fashion (threshold, or time, or both or something).  Once purged from memory, no way to get older data.
3. provide seemless way to read from memory / database if user requests old data, i imagine this work being the hardest.
4. provide some caching mechanism for recently read data from database

**Comment by grondo** [comment]
I would think #3 and #4 are not all that important at this time and can be delayed until much later.

For now the most important thing is to get the database going so the fairshare/accounting project can make progress.

Then the second priority would be to purge data from job-info (and possibly kvs) so that we don't run into resource exhaustion. The purge should be triggered only once data is safely in the database (i.e. not necessarily time or threshold based)

If it were me, I would treat the database as a completely separate service from job-info (at least for now).  It may be that trying to tie the two too tightly together is making the design more complex than it needs to be.

**Comment by garlick** [comment]
> If it were me, I would treat the database as a completely separate service from job-info (at least for now). It may be that trying to tie the two too tightly together is making the design more complex than it needs to be.

It might be useful to design this so the implementation can be  changed without major surgery, for example, down the road maybe we'll want to use an external database rather than embedded sqlite.  (Another argument for a separate module abstracted behind a message interface)

**Comment by grondo** [comment]
Another point I meant to mention: I believe Slurm purges jobs from memory after some short interval (500s by default), and if you want to see older job information you have to go to sacct/sqlog. It will probably be fine to have a similar scheme for Flux at first.
Then eventually offer a way to seamlessly list and query current + older jobs. This could even be done from a frontend command instead of embedded into a service (e.g. this is what we did with sqlog, it just queried two sources)

**Comment by chu11** [comment]
> If it were me, I would treat the database as a completely separate service from job-info (at least for now). It may be that trying to tie the two too tightly together is making the design more complex than it needs to be.

> It might be useful to design this so the implementation can be changed without major surgery, for example, down the road maybe we'll want to use an external database rather than embedded sqlite. (Another argument for a separate module abstracted behind a message interface)

Yeah, something that would be good / extensible is something I've been mulling over.  Short term, perhaps having an "internal service" in the job-info module that will simply get services using the `job-info.list-inactive` + `since` and stick em in some database is more than enough.

I believe we have to store the eventlog, R, and jobspec to fully recover data at the moment.  Not sure if we want to store the entire jobspec or only portions of it, as the jobspec can get large.

**Comment by garlick** [comment]
> I believe we have to store the eventlog, R, and jobspec to fully recover data at the moment. Not sure if we want to store the entire jobspec or only portions of it, as the jobspec can get large.

I think at one point we discussed aggregating the entire KVS space for a job into one JSON object. 
 Would that make sense here, as a column in the table so nothing is lost if we prune the KVS space?

**Comment by chu11** [comment]
> I think at one point we discussed aggregating the entire KVS space for a job into one JSON object.
Would that make sense here, as a column in the table so nothing is lost if we prune the KVS space?

Well, to make sure we don't lose anything it's the best idea.  But there's the tradeoff of fair-share having to parse too much here.

**Comment by chu11** [comment]
I just realized a big problem.  If I want to save off `jobspec`, `R`, or full `eventlog` in the database, we currently don't cache the full `jobspec`, `R`, or `eventlog` in memory.

It wouldn't be hard to add this, but that's a lot of data to cache in memory.  I guess short term it's fine to save it off?

**Comment by grondo** [comment]
> I just realized a big problem. If I want to save off jobspec, R, or full eventlog in the database, we currently don't cache the full jobspec, R, or eventlog in memory.

Does that matter, since we're talking about the database as completely separate service from job-info? The db service can fetch info directly from the kvs and create the job record directly without involving job-info.

**Comment by chu11** [comment]
> Does that matter, since we're talking about the database as completely separate service from job-info? The db service can fetch info directly from the kvs and create the job record directly without involving job-info.

I hadn't thought of getting the data from the KVS, I wanted to limit the number of lookup queries.  I "have to" call the job-info module to get a list of the inactive jobs and the job-info module has already retrieved eventlog, R, and jobspec anyways.  But perhaps going to the KVS is the best thing to do.

**Comment by chu11** [comment]
so just want to make sure that we're on the same page, what I'm thinking of doing

- separate flux-core module
- periodically retrieve new inactive job ids using `job-info.list-inactive` service
- get "raw" data from KVS (minimally eventlog, jobspec, and R), potentially put together into a single json object.
- stuff into db (round 1, I will likely just store into a text file)

**Comment by grondo** [comment]
For what you've described above, I  would think you could prototype with a python script instead of writing a module.

**Comment by cmoussa1** [comment]
> For what you've described above, I would think you could prototype with a python script instead of writing a module.

IMO Python has a pretty nice SQLite3 API that I've been using for the user accounting DB - https://docs.python.org/3/library/sqlite3.html

**Comment by chu11** [comment]
Well I ended up writing it in C, mostly b/c I already started it in C :P, but it ended there are some Python bindings that are missing too.  I have something really basic that "works", just writing `id`, `t_inactive`, `eventlog`, `jobspec`, and `R` to a file, newline separated (the `eventlog` has newlines in it, so need to encode that in some way).  Module just reads from `job-info.list-inactive` some configured period of seconds.  Then on re-load of the module, it will find the largest `t_inactive` in the file so it doesn't inadvertently read more data from `job-info.list-inactive`.

Some subtleties found during this prototyping:
- Generally speaking, configuration?  In the past we were doing module command line options.  Are we moving to TOML config files now for all modules?
- Where to store this file?  I assume we'll need some equivalent config to `content-backing.path`, since the old `persist` configs are gone.
- The initial call to get all inactive job IDs can be expensive, so working on #2778 (job-info list RPC restart) may be important to do.
- After getting a list of newly inactive job ids, it is a bit irritating to get the core data from the KVS
  - Since I want to store "eventlog", "jobspec", and "R" this would require three KVS lookups for each jobid
    - That can be a lot of simultaneous lookups in the KVS
  - Going through the job-info module's lookup service is nicer, but that's an extra RPC thrown in, and still only 1 jobid at a time.  Would be nice to have a "batch" lookup of some sort.
  - Another annoying thing, when you send a `job-info.lookup` request, the job id you requested isn't returned.  Not the end of the world, but annoying to have to do a `flux_future_aux_set/get()`.
- A subtle thing I realized, assuming you just want to append the results from a list-inactive lookup, the data in the file/database is not sorted based on recent completion.  This is b/c list-inactive returns job in most-recently completed first.  So if we care about the data being sorted in some way, we have to sort before appending.  I assume not a big deal.

None of the above is a deal breaker to getting some proto-type up and running.  I'm thinking I'll just get the trivial "file" implementation working (it's stupid, w/ lots of hacks), mostly just so we can get the skeleton of the module up and some tests of some sort.
